### PR TITLE
Model Output Visualization for PyTorch Models

### PR DESCRIPTION
**Pull request by n-poulsen**
This pull request implements methods to visualize maps output by PyTorch models (scoremaps, location refinement fields and part-affinity fields) as they are for TensorFlow models.

## API Methods Implemented for the PyTorch Engine

- `extract_maps`
- `visualize_scoremaps`
- `visualize_locrefs`
- `visualize_paf`
- `extract_save_all_maps`

## UX Improvements

- only start training pose models when `epochs > 0`
- `get_model_snapshots`: when no snapshots were found, the snapshot manager's snapshot prefix was obtained incorrectly
- fine-tuning SuperAnimal detectors: added `key_metric` so the best detector si saved

### CODE REVIEW

**Comment by MMathisLab** [comment]
@maximpavliv please test; I won't have time in the next days.

File: deeplabcut/compat.py [hunk 1]
```diff
@@ -1219,23 +1242,27 @@ def extract_maps(
...
     raise NotImplementedError(f"This function is not implemented for {engine}")
 
 
-def visualize_scoremaps(
-    image: np.ndarray, scmap: np.ndarray, engine: Engine = DEFAULT_ENGINE,
-):
+def visualize_scoremaps(image: np.ndarray, scmap: np.ndarray):
     """
     This function is only implemented for tensorflow models/shuffles, and will throw
     an error if called with a PyTorch shuffle.
```
Comment (maximpavliv) [inline deeplabcut/compat.py]: Remove the warning, and maybe a little description instead would be great

File: deeplabcut/compat.py [hunk 2]
```diff
@@ -1245,35 +1272,27 @@ def visualize_locrefs(
...
 def visualize_paf(
     image: np.ndarray,
     paf: np.ndarray,
     step: int = 5,
     colors: list | None = None,
-    engine: Engine = DEFAULT_ENGINE,
 ):
     """
     This function is only implemented for tensorflow models/shuffles, and will throw
     an error if called with a PyTorch shuffle.
```
Comment (maximpavliv) [inline deeplabcut/compat.py]: Same here

File: deeplabcut/compat.py [hunk 3]
```diff
@@ -1283,11 +1302,14 @@ def extract_save_all_maps(
...
+    device: str | None = None,
     rescale: bool = False,
     Indices: list[int] | None = None,
     modelprefix: str = "",
     dest_folder: str = None,
+    snapshot_index: int | str | None = None,
+    detector_snapshot_index: int | str | None = None,
     engine: Engine | None = None,
 ):
     """
```
Comment (maximpavliv) [inline deeplabcut/compat.py:1336]: In the docstring below, remove _"This function is only implemented for tensorflow models/shuffles, and will throw an error if called with a PyTorch shuffle."_

File: deeplabcut/compat.py [hunk 4]
```diff
@@ -1158,6 +1170,7 @@ def extract_maps(
     shuffle: int = 0,
     trainingsetindex: int = 0,
     gputouse: int | None = None,
+    device: str | None = None,
     rescale: bool = False,
     Indices: list[int] | None = None,
     modelprefix: str = "",
```
Comment (maximpavliv) [inline deeplabcut/compat.py:1176]: In the docstring below, remove _"This function is only implemented for tensorflow models/shuffles, and will throw an error if called with a PyTorch shuffle."_

File: deeplabcut/pose_estimation_pytorch/runners/inference.py [hunk 5]
```diff
@@ -128,7 +128,7 @@ def _prepare_inputs(
         """
```
Comment (maximpavliv) [inline deeplabcut/pose_estimation_pytorch/runners/inference.py:128]: Add Path to the argument type suggestions

File: docs/pytorch/user_guide.md [hunk 6]
```diff
@@ -69,9 +69,9 @@ print(available_models())
 
 ### Development State and Road Map 🚧
 
-The table below describes the DeepLabCut API methods that have been implemented,
-as well as indications which options are not yet implemented, and which parameters
-are not valid for the DLC 3.0 API.
+The table below describes the DeepLabCut API methods that have been implemented for the
+PyTorch, as well as indications which options are not yet implemented, and which
```
Comment (maximpavliv) [inline docs/pytorch/user_guide.md]:
```suggestion
PyTorch API, as well as indications which options are not yet implemented, and which
```
I think that there is a word missing, API / version / release ?

**Review by maximpavliv — CHANGES_REQUESTED** [review]
Some minor corrections :)

**Comment by n-poulsen** [comment]
@maximpavliv good catches, thanks! I implemented the fixes from your suggestions

File: deeplabcut/compat.py [hunk 7]
```diff
@@ -1219,23 +1242,27 @@ def extract_maps(
             Indices=Indices,
             modelprefix=modelprefix,
         )
+    elif engine == Engine.PYTORCH:
+        from deeplabcut.pose_estimation_pytorch import extract_maps
+        extract_maps(
```
Comment (maximpavliv) [inline deeplabcut/compat.py]:
```suggestion
        return extract_maps(
```
Otherwise it ends in the `raise NotImplementedError(...)`

**Review by maximpavliv — CHANGES_REQUESTED** [review]

**Review by maximpavliv — APPROVED** [review]
